apps/playwright-browser-tunnel/src/tunneledBrowserConnection.ts

```diff
@@ -0,0 +1,299 @@
+          const bName: string | null = parsed.searchParams.get('browser');
+          if (bName && ['chromium', 'firefox', 'webkit'].includes(bName)) {
+            browserName = bName as BrowserNames;
+          }
+          const launchOptionsParam: string | null = parsed.searchParams.get('launchOptions');
+          if (launchOptionsParam) {
+            try {
+              launchOptions = JSON.parse(launchOptionsParam);
+            } catch (e) {
+              console.error('Invalid launchOptions JSON provided');
+            }
+          }
+        }
+      } catch (e) {
+        console.error(`Error parsing local connection query params: ${e}`);
+      }
+
+      if (!browserName) {
+        console.error('browser query param required (chromium|firefox|webkit)');
+        localWs.close();
+        return;
+      }
+      if (!launchOptions) {
+        launchOptions = {} as LaunchOptions; // default empty if not provided
+      }
+
+      maybeSendHandshake();
+
+      localWs.on('message', (message) => {
+        if (handshakeAck && remoteSocket && remoteSocket.readyState === WebSocket.OPEN) {
+          remoteSocket.send(message);
+        } else {
+          // Buffer until handshakeAck to avoid losing early protocol messages from Playwright
+          bufferedLocalMessages.push(message);
+        }
+      });
+      localWs.on('close', () => console.log('Local client websocket closed'));
+      localWs.on('error', (err) => console.error(`Local client websocket error: ${err}`));
+    });
+
+    // Resolve immediately so caller can initiate local connection with query params (handshake completes later)
+    resolve({
+      remoteEndpoint: localProxyWsEndpoint,
+      [Symbol.dispose]() {
+        try {
+          remoteWsServer.close();
+        } catch {
+          // ignore errors during remote WebSocket server shutdown
+        }
+        try {
+          localProxyWs.close();
+        } catch {
+          // ignore errors during local proxy WebSocket server shutdown
+        }
+      },
+      // eslint-disable-next-line promise/param-names
+      closePromise: new Promise<void>((resolve2) => {
+        remoteWsServer.on('close', () => {
+          resolve2();
+        });
+      })
+    });
+  });
+}
+
+/**
+ * Disposable handle returned by {@link tunneledBrowser}.
+ * @alpha
+ */
+export interface IDisposableTunneledBrowser {
+  browser: Browser;
+  [Symbol.asyncDispose]: () => Promise<void>;
+}
+
+/**
+ * Creates a Playwright Browser instance connected via a tunneled WebSocket connection.
+ * @alpha
+ */
+export async function tunneledBrowser(
+  browserName: BrowserNames,
+  launchOptions: LaunchOptions
+): Promise<IDisposableTunneledBrowser> {
+  // Establish the tunnel first (remoteEndpoint here refers to local proxy endpoint for connect())
+  using connection: IDisposableTunneledBrowserConnection = await tunneledBrowserConnection();
```

Comment:
The `connection` will be disposed as soon as the `return` statement resolves; it seems more likely that `connection` should get disposed in the returned `[Symbol.asyncDispose]` method.